tools/internal/parser/parser_test.go

```diff
@@ -0,0 +1,777 @@
+package parser
+
+import (
+	"bytes"
+	"cmp"
+	"net/mail"
+	"net/url"
+	"os"
+	"slices"
+	"strings"
+	"testing"
+
+	diff "github.com/google/go-cmp/cmp"
+)
+
+// TestParser runs a battery of synthetic parse and validation tests.
+func TestParser(t *testing.T) {
+	// These test cases have a fair amount of repetition in them,
+	// since both errors and suffix blocks contain repeated nestings
+	// of blocks and Source objects. While it's tempting to try and
+	// reduce duplication through clever code, you are encouraged to
+	// resist the urge.
+	//
+	// Each test case is quite verbose, but being laid out with
+	// minimal indirection makes it easier to inspect and debug when a
+	// failure happens.
+
+	tests := []struct {
+		name               string
+		psl                string
+		downgradeToWarning func(error) bool
+		want               File
+	}{
+		{
+			name: "empty",
+			psl:  "",
+			want: File{},
+		},
+
+		{
+			name: "just_comments",
+			psl: dedent(`
+			  // This is an empty PSL file.
+
+			  // Here is a second comment.
+			`),
+			want: File{
+				Blocks: []Block{
```

Comment:
Should this File contain errors because the sections are missing?

Reply:
Currently the parser does not require the presence of specific sections, it only requires that the start/end markers are paired correctly.

Added this validation to my TODOs.